src/cs/Ssh/SshSession.cs

```diff
@@ -171,6 +174,16 @@ internal SshSession(
 	/// </remarks>
 	public event EventHandler<EventArgs>? Disconnected;
 
+	/// <summary>
+	/// Event is raised when a keep-alive request is sent but no response is received
+	/// </summary>
+	public event EventHandler<SshKeepAliveEventArgs>? KeepAliveRequestFailed;
+
+	/// <summary>
+	/// Internal event used for testing purposes.
+	/// </summary>
+	internal event EventHandler<EventArgs>? KeepAliveRequestReceived;
```

Comment:
Instead of this internal event just for testing, I'd suggest using trace events:
1. Define a new event ID in `SshTraceEventIds`.
2. When a keep-alive request is received, trace an event with that ID instead of raising `KeepAliveRequestReceived`.
3. In test cases, listen to trace events check and for that ID. See the `TraceChannelData()` test case for an example of this testing approach. You can re-use `TestTraceListener` - make it a top-level class in the test assembly.

Consider also adding trace event IDs for when a keep-alive message is sent and when a keep-alive reply was not received. Those may help with further testing.

Reply:
Thanks for pointing about the `TestTraceListener`, I was able to test those scenarios with that.